homeworks/JohannesMerkt/5-remesh/src/geometry/halfedge.ts

```diff
@@ -0,0 +1,460 @@
+        v2Halfedges.set(halfedge.idx, halfedge);
+      });
+
+      // create a map of all edges that need to have their error updated after modification
+      const edges2Update = new Map<number, Edge>();
+      v1.halfedges(halfedge => {
+        if (halfedge.edge) {
+          edges2Update.set(halfedge.edge.idx, halfedge.edge);
+          halfedge.twin!.vert!.halfedges(he => {
+            if (he.edge) {
+              edges2Update.set(he.edge.idx, he.edge);
+            }
+          });
+        }
+      });
+      v2.halfedges(halfedge => {
+        if (halfedge.edge) {
+          edges2Update.set(halfedge.edge.idx, halfedge.edge);
+          halfedge.twin!.vert!.halfedges(he => {
+            if (he.edge) {
+              edges2Update.set(he.edge.idx, he.edge);
+            }
+          });
+        }
+      });
+      v3.halfedges(halfedge => {
+        if (halfedge.edge) {
+          edges2Update.set(halfedge.edge.idx, halfedge.edge);
+          halfedge.twin!.vert!.halfedges(he => {
+            if (he.edge) {
+              edges2Update.set(he.edge.idx, he.edge);
+            }
+          });
+        }
+      });
+      v4.halfedges(halfedge => {
+        if (halfedge.edge) {
+          edges2Update.set(halfedge.edge.idx, halfedge.edge);
+          halfedge.twin!.vert!.halfedges(he => {
+            if (he.edge) {
+              edges2Update.set(he.edge.idx, he.edge);
+            }
+          });
+        }
+      });
+
+      // make sure reused elements reference elements that are not deleted
+      e13.halfedge = h31;
+      e14.halfedge = h14;
+      v1.halfedge = h14;
+      v3.halfedge = h31;
+      v4.halfedge = h42;
+      h23.vert = v1;
+      v2Halfedges.forEach(halfedge => {
+        halfedge.vert = v1;
+      });
+      // update edges
+      h42.edge = e14;
+      h23.edge = e13;
+      // update twin relationship
+      h14.twin = h42;
+      h42.twin = h14;
+      h31.twin = h23;
+      h23.twin = h31;
+      // update v1 vert position
+      v1.pos = e12.bestVertex();
+      // remove edge, face and halfedges that are collapsed
+      e12.removed = true;
+      e24.removed = true;
+      e23.removed = true;
+      this.edges[e12.idx] = null;
+      this.edges[e24.idx] = null;
+      this.edges[e23.idx] = null;
+      this.verts[v2.idx] = null;
+      f1.removed = true;
+      f2.removed = true;
+      this.faces[f1.idx] = null;
+      this.faces[f2.idx] = null;
+      this.halfedges[h12.idx] = null;
+      this.halfedges[h21.idx] = null;
+      this.halfedges[h24.idx] = null;
+      this.halfedges[h41.idx] = null;
+      this.halfedges[h32.idx] = null;
+      this.halfedges[h13.idx] = null;
+
+      currentFaceCount = currentFaceCount - 2;
+      // update error value of edges
+      edges2Update.forEach(edge => {
+        if (!edge.removed) {
+          edgeErrorMap.set(edge.idx, edge.error());
+        }
+      });
+      edgeQueue.sort((a, b) => edgeErrorMap.get(b)! - edgeErrorMap.get(a)!);
```

Comment:
Sort introduces another nlog(n) complexity?

Reply:
This sort is likely the most timeconsuming. I didnt know how to make this part faster. Do you have any idea?